lib/kino_maplibre/map_cell.ex

```diff
@@ -0,0 +1,326 @@
+
+    ml_alias = ml_alias(env)
+    send(pid, {:scan_binding_result, source_variables, ml_alias})
+  end
+
+  @impl true
+  def handle_connect(ctx) do
+    payload = %{
+      root_fields: ctx.assigns.root_fields,
+      sources: ctx.assigns.sources,
+      layers: ctx.assigns.layers,
+      source_variables: ctx.assigns.source_variables,
+      missing_dep: ctx.assigns.missing_dep
+    }
+
+    {:ok, payload, ctx}
+  end
+
+  @impl true
+  def handle_info({:scan_binding_result, source_variables, ml_alias}, ctx) do
+    ctx = assign(ctx, ml_alias: ml_alias, source_variables: source_variables)
+    broadcast_event(ctx, "set_source_variables", %{"source_variables" => source_variables})
+    {:noreply, ctx}
+  end
+
+  @impl true
+  def handle_event("update_field", %{"field" => field, "value" => value, "idx" => nil}, ctx) do
+    parsed_value = parse_value(field, value)
+    ctx = update(ctx, :root_fields, &Map.put(&1, field, parsed_value))
+    broadcast_event(ctx, "update_root", %{"fields" => %{field => parsed_value}})
+    {:noreply, ctx}
+  end
+
+  def handle_event(
+        "update_field",
+        %{"field" => "source_type" <> _, "value" => value, "idx" => idx},
+        ctx
+      ) do
+    source = get_in(ctx.assigns.sources, [Access.at(idx)])
+    source_data = if value == "variable", do: List.first(ctx.assigns.source_variables).variable
+    updated_source = %{source | "source_type" => value, "source_data" => source_data}
+    updated_sources = List.replace_at(ctx.assigns.sources, idx, updated_source)
+    ctx = update_in(ctx.assigns, fn assigns -> Map.put(assigns, :sources, updated_sources) end)
+
+    broadcast_event(ctx, "update_source", %{
+      "idx" => idx,
+      "fields" => %{"source_type" => value, "source_data" => source_data}
+    })
+
+    {:noreply, ctx}
+  end
+
+  def handle_event("update_field", %{"field" => field, "value" => value, "idx" => idx}, ctx)
+      when field in @source_fields do
+    parsed_value = parse_value(field, value)
+    updated_sources = put_in(ctx.assigns.sources, [Access.at(idx), field], parsed_value)
+    ctx = update_in(ctx.assigns, fn assigns -> Map.put(assigns, :sources, updated_sources) end)
+    broadcast_event(ctx, "update_source", %{"idx" => idx, "fields" => %{field => parsed_value}})
+
+    {:noreply, ctx}
+  end
+
+  def handle_event("update_field", %{"field" => field, "value" => value, "idx" => idx}, ctx) do
+    parsed_value = parse_value(field, value)
+    updated_layers = put_in(ctx.assigns.layers, [Access.at(idx), field], parsed_value)
+    ctx = update_in(ctx.assigns, fn assigns -> Map.put(assigns, :layers, updated_layers) end)
+    broadcast_event(ctx, "update_layer", %{"idx" => idx, "fields" => %{field => parsed_value}})
+
+    {:noreply, ctx}
+  end
+
+  def handle_event("add_source", _, ctx) do
+    updated_sources = ctx.assigns.sources ++ empty_source()
+    ctx = update_in(ctx.assigns, fn assigns -> Map.put(assigns, :sources, updated_sources) end)
```

Comment:
Could this `update_in` be replaced by the update syntax of maps?

```suggestion
    ctx = %{ctx | assigns: %{ctx.assigns | sources: updated_sources}}
```

Reply:
I would also recommend this for other events in this file. Maybe creating a helper can be good for this.

Reply:
Nice! It worked! About the helper... There's already a helper on the Kino side, but it doesn't handle nested data well. I think it would be the best place for the helper, so all the smart cells could use it. I'll see! Thanks ❤️